base/strings/util.jl

```diff
@@ -513,6 +513,101 @@ function rpad(
     r == 0 ? stringfn(s, p^q) : stringfn(s, p^q, first(p, r))
 end
 
+"""
+    rtrunc(str::AbstractString, maxwidth::Int, replace_str::AbstractString = "…")
+
+Truncate `str` to at most `maxwidth` columns (as estimated by [`textwidth`](@ref)), replacing the last characters
+with `replacement` if necessary. The default replacement string is "…".
+
+# Examples
+```jldoctest
+julia> s = rtrunc("🍕🍕 I love 🍕", 10)
+"🍕🍕 I lo…"
+
+julia> textwidth(s)
+10
+
+julia> rtrunc("foo", 3)
+"foo"
+```
+
+!!! compat "Julia 1.12"
+    This function was added in Julia 1.12.
+
+See also [`ltrunc`](@ref) and [`ctrunc`](@ref).
+"""
+rtrunc(str::AbstractString, maxwidth::Int, replacement::Union{AbstractString,Char} = '…') = struncate(str, maxwidth, replacement, :right)
+
+"""
+    ltrunc(str::AbstractString, maxwidth::Int, replace_str::AbstractString = "…")
+
+Truncate `str` to at most `maxwidth` columns (as estimated by [`textwidth`](@ref)), replacing the first characters
+with `replacement` if necessary. The default replacement string is "…".
+
+# Examples
+```jldoctest
+julia> s = ltrunc("🍕🍕 I love 🍕", 10)
+"…I love 🍕"
+
+julia> textwidth(s)
+10
+
+julia> ltrunc("foo", 3)
+"foo"
+```
+
+!!! compat "Julia 1.12"
+    This function was added in Julia 1.12.
+
+See also [`rtrunc`](@ref) and [`ctrunc`](@ref).
+"""
+ltrunc(str::AbstractString, maxwidth::Int, replacement::Union{AbstractString,Char} = '…') = struncate(str, maxwidth, replacement, :left)
+
+"""
+    ctrunc(str::AbstractString, maxwidth::Int, replacement::Union{AbstractString,Char} = '…'; prefer_left::Bool = true)
+
+Truncate `str` to at most `maxwidth` columns (as estimated by [`textwidth`](@ref)), replacing the middle characters
+with `replacement` if necessary. The default replacement string is "…". By default, the truncation
+prefers keeping chars on the left, but this can be changed by setting `prefer_left` to `false`.
+
+# Examples
+```jldoctest
+julia> s = ctrunc("🍕🍕 I love 🍕", 10)
+"🍕🍕 …e 🍕"
+
+julia> textwidth(s)
+10
+
+julia> ctrunc("foo", 3)
+"foo"
+```
+
+!!! compat "Julia 1.12"
+    This function was added in Julia 1.12.
+
+See also [`ltrunc`](@ref) and [`rtrunc`](@ref).
+"""
+ctrunc(str::AbstractString, maxwidth::Int, replacement::Union{AbstractString,Char} = '…'; prefer_left::Bool = true) =
+    struncate(str, maxwidth, replacement, :center, prefer_left)
+
+function struncate(str::AbstractString, maxwidth::Int, replacement::Union{AbstractString,Char}, mode::Symbol = :center, prefer_left::Bool = true)
```

Comment:
```suggestion
function _struncate(str::AbstractString, maxwidth::Int, replacement::Union{AbstractString,Char}, ::Val{mode}, prefer_left::Bool = true) where {mode}
```
use an underscore to indicate that this is an internal method, and use `Val` so that the `mode` checks can be elided by the compiler?

Reply:
Now that we have `public` we don't need to mark internals by name convention?

Val sounds good. I didn't appreciate why before.

Reply:
Why `Val` and not e.g. a keyword arg?

Reply:
> Now that we have `public` we don't need to mark internals by name convention?

because the REPL autocompletes and even gives hints for private names: #51327

Reply:
Regarding `mode`, IMO something like an enum would be preferable to just `Symbol`, due to being typo-proof. But maybe that's not important if `struncate` is to stay private.

Reply:
> Regarding mode, IMO something like an enum would be preferable to just Symbol, due to being typo-proof.

I want to make sure that the `mode == ...` conditionals are evaluated at compile time.   In principle this could also happen with enum or `Symbol` arguments if we ensure that the function is inlined.

Another possibility would be singleton types.  But that may be overkill for an internal `_struncate` method.

Reply:
> I want to make sure that the mode == ... conditionals are evaluated at compile time. 

This should be backed up by some benchmark in my opinion. A single pointer equality check seems to me insignificant from the work of the function.

In principle this could also happen with enum or Symbol arguments if we ensure that the function is inlined.

I don't think constant propagation requires inlining. And there is also `@constprop :aggressive`.

`Val` is never really a good idea to make an API upon (any longer).

Reply:
> This should be backed up by some benchmark in my opinion. A single pointer equality check seems to me insignificant from the work of the function.

I tried this with
```jl
s = randstring(10^5)
@btime struncate($s, 100, '…', :center);
```
and I got about an 8% speedup by hard-coding the `mode` checks.  (This is *after* doing the `total_width` optimization noted below — otherwise, it spends all its time on the `textwidth(str)` check.)
```

Reply:
I get another 60% speedup by replacing `textwidth(::Char)` with an optimized version that has a fast path for ASCII characters.  We might want to do this in a separate PR:
```jl
const textwidth_ascii = Cint[textwidth(Char(i)) for i = 0:127]
function textwidth2(c::AbstractChar)
    b = bswap(reinterpret(UInt32, c))
    b < 0x80 && @inbounds return Int(textwidth_ascii[b+1]) # ASCII fast path
    Base.ismalformed(c) && return 1
    Int(ccall(:utf8proc_charwidth, Cint, (UInt32,), c))
end
```

Reply:
Using `Val` I do see reduction from
```
julia> @b ctrunc("dsadsafsadsadsafdsad", 10)
150.062 ns (4 allocs: 112 bytes)
```
to
```
136.986 ns (4 allocs: 112 bytes)
```
That's without #55398